File: ios/LaneShadow/Views/Atoms/Chip.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Chip Component

/**
 * Chip component
 *
 * Following RN wrapper API from react-native/components/ui/chip.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/atoms/Chip.md
 *
 * ## Design Tokens Used
 * - Padding: `theme.space.md` (12) horizontal, 6 vertical
 * - Radius: `theme.radius.full` (9999)
 * - Border: `theme.borderWidth.thin` (1)
 * - Colors: `theme.colors.primary`, `theme.colors.border`, `theme.colors.onSurface`, `theme.colors.muted`
 * - Typography: 13pt medium (between label.sm 12 and body.sm 14)
 * - Icon: 16pt, `theme.space.xs` (4) gap
 * - Opacity: 12% (selected background), 40% (selected border)
 *
 * ## Parameters
 * - Parameters:
 *   - label: Text label to display
 *   - selected: Whether the chip is selected (default: false)
 *   - onPress: Action callback when chip is pressed
 *   - icon: Optional icon view to display
 *   - testID: Test identifier for UI testing
 */
public struct LSChip: View {
    @Environment(\.theme) private var theme
    @State private var isPressed = false

    private let label: String
    private let selected: Bool
    private let onPress: (() -> Void)?
    private let icon: (() -> AnyView)?
    private let testID: String?

    public init(
        _ label: String,
        selected: Bool = false,
        onPress: (() -> Void)? = nil,
        icon: (() -> AnyView)? = nil,
        testID: String? = nil
    ) {
        self.label = label
        self.selected = selected
        self.onPress = onPress
        self.icon = icon
        self.testID = testID
    }

    // MARK: - Color Computed Properties

    private var backgroundColor: Color {
        if selected {
            return theme.colors.primary.default.opacity(0.12)
        }

        if isPressed {
            return theme.colors.muted.pressed ?? theme.colors.muted.default
        }

        return .clear
    }

    private var borderColor: Color {
        if selected {
            return theme.colors.primary.default.opacity(0.4)
        }

        return theme.colors.border.default
    }

    private var textColor: Color {
        if selected {
            return theme.colors.primary.default
        }

        return theme.colors.onSurface.default
    }

    private var iconColor: Color {
        if selected {
            return theme.colors.primary.default
        }

        return theme.colors.onSurface.muted ?? theme.colors.onSurface.default
    }

    // MARK: - Body

    public var body: some View {
        Group {
            if let icon {
                chipContent(
                    label: AnyView(
                        HStack(spacing: theme.space.xs) {
                            icon()
                                .foregroundStyle(iconColor)
                            Text(label)
                        }
                    )
                )
            } else {
                chipContent(
                    label: AnyView(
                        Text(label)
                    )
                )
            }
        }
        .accessibilityAddTraits(.isButton)
        .accessibilityAddTraits(selected ? .isSelected : [])
        .accessibilityIdentifier(testID ?? "chip")
    }

    // MARK: - Chip Content

    @ViewBuilder
    private func chipContent(label: AnyView) -> some View {
        let content = label
            .font(.system(size: 13, weight: .medium))
            .foregroundStyle(textColor)
            .padding(.horizontal, theme.space.md)
            .padding(.vertical, 6)
            .background(backgroundColor)
            .clipShape(Capsule())
            .overlay(
                Capsule()
                    .stroke(borderColor, lineWidth: theme.borderWidth.thin)
            )

        if let onPress = onPress {
            Button(action: onPress) {
                content
            }
            .buttonStyle(.plain)
            .simultaneousGesture(
                DragGesture(minimumDistance: 0)
                    .onChanged { _ in
                        if !selected {
                            isPressed = true
                        }
                    }
                    .onEnded { _ in
                        isPressed = false
                    }
            )
        } else {
            content
        }
    }
}

File: android/app/src/main/java/com/laneshadow/ui/components/atoms/Chip.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.PressInteraction
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * Chip atom component
 *
 * Selectable chip component with icon support.
 * Following RN wrapper API from react-native/components/ui/chip.tsx
 *
 * Style properties from matrix:
 * - Padding: horizontal 12dp (space.md), vertical 6dp
 * - Radius: full (9999dp / CircleShape)
 * - Border: 1dp
 * - Font: 13sp, Medium weight, 18sp line height
 * - Icon: 16dp size
 * - Icon gap: 4dp (space.xs)
 *
 * @param label Text label for the chip
 * @param selected Whether the chip is selected
 * @param onPress Callback when chip is pressed (null makes chip non-interactive)
 * @param icon Optional icon composable to display
 * @param testID Test ID for UI testing
 */
@Composable
fun Chip(
    label: String,
    selected: Boolean = false,
    onPress: (() -> Unit)? = null,
    icon: @Composable (() -> Unit)? = null,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Track pressed state for visual feedback
    var isPressed by remember { mutableStateOf(false) }
    val interactionSource = remember { MutableInteractionSource() }

    LaunchedEffect(interactionSource) {
        interactionSource.interactions.collect { interaction ->
            when (interaction) {
                is PressInteraction.Press -> isPressed = true
                is PressInteraction.Release -> isPressed = false
                is PressInteraction.Cancel -> isPressed = false
            }
        }
    }

    // Background color based on state
    val backgroundColor: Color = when {
        selected -> theme.colors.primary.default.copy(alpha = 0.12f)
        isPressed && !selected -> theme.colors.muted.pressed ?: theme.colors.muted.default
        else -> Color.Transparent
    }

    // Border color based on state
    val borderColor = when {
        selected -> theme.colors.primary.default.copy(alpha = 0.4f)
        else -> theme.colors.border.default
    }

    // Text color based on state
    val textColor = if (selected) {
        theme.colors.primary.default
    } else {
        theme.colors.onSurface.default
    }

    // Icon color based on state (for icon rendering)
    val iconColor = if (selected) {
        theme.colors.primary.default
    } else {
        theme.colors.onSurface.default.copy(alpha = 0.6f)
    }

    // Build modifier with semantics and click handling
    val chipModifier = Modifier
        .semantics {
            role = Role.Checkbox
            this.selected = selected
        }
        .then(
            if (onPress != null) {
                Modifier.clickable(
                    onClick = onPress,
                    interactionSource = interactionSource,
                    indication = null,
                )
            } else {
                Modifier
            }
        )

    Row(
        modifier = chipModifier
            .background(
                color = backgroundColor,
                shape = CircleShape,
            )
            .border(
                border = BorderStroke(1.dp, borderColor),
                shape = CircleShape,
            )
            .padding(
                horizontal = 12.dp,
                vertical = 6.dp,
            ),
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        if (icon != null) {
            icon()
        }

        Text(
            text = label,
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
            lineHeight = 18.sp,
            color = textColor,
        )
    }
}
